byte-based splitting for mse query blocks

albertobastos opened the pull request:
So far query blocks sent through gRPC by the Multi Stage Engine follow a row-based splitting strategy which is not ideal, given that a single row could be larger than the maximum block size allowed.

This PR creates a new splitting strategy on top of that based purely on the block bytes payload. Both strategies can live simultaneously, although it is expected that byte-based splitter will take over and row-based splitter can be eventually removed.

gortiz commented on pinot-common/src/main/java/org/apache/pinot/common/datablock/DataBlockUtils.java:
```diff
@@ -230,6 +230,28 @@ public static ByteString toByteString(DataBlock dataBlock)
+      throws IOException {
+    List<ByteBuffer> bytes = dataBlock.serialize();
+    if (bytes.isEmpty()) {
+      return List.of(ByteString.EMPTY);
```
Why not just an empty list? In case this is actually necessary, I think it should be added in this method javadoc

gortiz commented on pinot-common/src/main/java/org/apache/pinot/common/datablock/DataBlockUtils.java:
```diff
@@ -230,6 +230,28 @@ public static ByteString toByteString(DataBlock dataBlock)
+      return List.of(ByteString.EMPTY);
+    }
+
+    List<ByteString> byteStrings = new ArrayList<>();
```
Nit: it may be worth calculating the initial size of the array list. I think an upper bound would be to calculate the total byte size (iterating over `bytes`) and then add the number of elements in `bytes` (as in the worst case scenario, we add an extra small ByteString for each page in `bytes`)

gortiz commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/GrpcSendingMailbox.java:
```diff
@@ -157,19 +167,28 @@ private StreamObserver<MailboxContent> getContentObserver() {
-      ByteString byteString = DataBlockUtils.toByteString(dataBlock);
-      int sizeInBytes = byteString.size();
+      // so far we ensure payload is not bigger than maxBlockSize/2, we can fine tune this later
+      List<ByteString> byteStrings = DataBlockUtils.toByteStrings(dataBlock, getMaxBlockSize() / 2);
```
`getMaxBlockSize()` won't change once the app is started, right? Do we need to read configs and parse strings to int once per block sent?

gortiz commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/channel/MailboxContentObserver.java:
```diff
@@ -58,10 +61,18 @@ public void onNext(MailboxContent mailboxContent) {
     if (_mailbox == null) {
       _mailbox = _mailboxService.getReceivingMailbox(mailboxId);
     }
+    if (_mailboxBuffers == null) {
+      _mailboxBuffers = new ArrayList<>();
+    }
```
nit: What is the advantage of this lazy initialization instead of doing it eagerly?

gortiz replied:
tip: ArrayLists do not allocate the array on construction unless you specify a custom size (see https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/java.base/share/classes/java/util/ArrayList.java#L166)

albertobastos replied:
Nice one. I always assumed it just used a default capacity, but that makes a lot more sense.

gortiz reviewed (commented):
Looks almost good to me. I think we can add some extra tests (ie DataBlockUtils.toByteStrings, but may be other methods/paths)

albertobastos reviewed (commented):
Thanks for the review @gortiz.

Updated with your suggestions and added a test for the `toByteStrings` method. I thought about creating a second test for the _returning empty list_ case, but `DataBlockUtils` is already assuming that serialization will return a `DataBuffer` with at least one byte, so didn't see the point on that,.

gortiz commented on pinot-common/src/main/java/org/apache/pinot/common/datablock/DataBlockUtils.java:
```diff
@@ -230,6 +230,28 @@ public static ByteString toByteString(DataBlock dataBlock)
+      } else {
+        current = current.concat(UnsafeByteOperations.unsafeWrap(bb));
+      }
+    }
```
Why would it need array copies? we just need to create views on the bytebuffer, right? The only issue is that it may be more difficult to calculate the size of the list where we are going to store these pages

albertobastos replied:
That took a while and some struggling using so many bytebuffer wrappers, but I believe we finally got it.

Changed the implementation so the split strategy is more optimal: it fills as many _maxByteStringSize_ `ByteString`s instances as needed, splitting the raw data among them. No buffer copies involved.

albertobastos commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/GrpcSendingMailbox.java:
```diff
@@ -229,4 +254,64 @@ public DataBlock visit(ErrorMseBlock block, List<DataBuffer> serializedStats) {
+  }
+
+  // polyfill because ByteBuffer.slice(pos, lim) is not available until Java 13
+  private static ByteBuffer sliceByteBuffer(ByteBuffer bb, int position, int limit) {
+    int oldPosition = bb.position();
+    int oldLimit = bb.limit();
+
+    try {
+      bb.position(position);
+      bb.limit(limit);
+      return bb.slice();
+    } finally {
+      bb.position(oldPosition);
+      bb.limit(oldLimit);
+    }
```
As we talked offline, it depends on what we want exactly to achieve during the splitting process.

If we want to optimize the amount of output chunks (filling chunks of `maxByteStringSize` until the very last chunk) it doesn't get much better than this. If we only want to bother splitting when a `ByteBuffer` exceeds the `maxByteStringSize` and we're ok with just pushing as-is the ones smaller than that, we could get rid of the slicing (although we will still need to update the buffer internal limit and position during the splitting).

I'm not convinced yet about changing it. Talking about performance, wouldn't say that updating limit and position cursors are too expensive (internally is just some simple assignations and comparisons). Talking about readibility... I believe current code is quite readable, to be honest 😅

gortiz replied:
Moving positions is not trivial (it checks boundaries each time), but the most expensive part here is actually slicing, which requires a nontrivial new allocation (5 ints, a long, 3 booleans and one reference to a null byte[]). Keep in mind that UnsafeByteOperations.unsafeWrap creates an additional slice under the hood.

gortiz commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/GrpcSendingMailbox.java:
```diff
@@ -52,24 +58,35 @@
 
+  private static final List<ByteString> EMPTY_BYTEBUFFER_LIST = Collections.emptyList();
+
+  private final PinotConfiguration _config;
```
Do you need the attribute? it looks like it is only used in the constructor

albertobastos replied:
Yeah, didn't remove it after moving some code to the constructor. Thanks for removing it.

yashmayya commented on pinot-query-runtime/src/test/java/org/apache/pinot/query/mailbox/GrpcSendingMailboxTest.java:
```diff
@@ -0,0 +1,172 @@
+    ByteBuffer expected = concatenateBuffers(input);
+
+    List<ByteString> output = GrpcSendingMailbox.toByteStrings(input, maxByteStringSize);
+    for (ByteString chunk: output) {
+      assertTrue(chunk.size() <= maxByteStringSize);
+    }
```
Minor suggestion to strengthen the assertion:
```suggestion
    for (ByteString chunk: output.subList(0, output.size() - 1)) {
      assertEquals(chunk.size(), maxByteStringSize);
    }
    assertTrue(output.get(output.size() - 1).size() <= maxByteStringSize);
```

yashmayya commented:
> This PR creates a new splitting strategy on top of that based purely on the block bytes payload. Both strategies can live simultaneously, although it is expected that byte-based splitter will take over and row-based splitter can be eventually removed.

@gortiz the `BlockSplitter` used in `BlockExchange` to split blocks based on rows doesn't serve any other purpose than trying to keep the gRPC message size under the max limit right? Any reason we can't get rid of that with this new strategy immediately? I'm not sure if I'm missing anything but it seems like unnecessary overhead now?

gortiz commented on pinot-common/src/main/java/org/apache/pinot/common/datablock/BaseDataBlock.java:
```diff
@@ -402,7 +403,14 @@ public List<ByteBuffer> serialize()
       _serialized = DataBlockUtils.serialize(this);
     }
-    return _serialized;
+    // Return a copy of the serialized data to avoid external modification.
+    List<ByteBuffer> copy = new ArrayList<>(_serialized.size());
+    for (ByteBuffer page: _serialized) {
+      ByteBuffer pageCopy = page.duplicate();
+      pageCopy.order(page.order());
+      copy.add(pageCopy);
+    }
+    return copy;
```
Exactly. And we only cache the serialization buffers because in the current code it is not easy to serialize blocks only once. It would be cleaner to serialize once at mailbox send operator and then send these bytes to each SendingMailbox, but there are sending mailboxes (ie InMemorySendingMailbox) that prefer the heap object, so we would need to refactor the code a bit.

Jackie-Jiang replied:
This is adding overhead to every single data block. Do we need to cache the serialized buffers? Are we ever serializing the same block twice?

gortiz commented on pinot-common/src/main/java/org/apache/pinot/common/datablock/DataBlockUtils.java:
```diff
@@ -252,7 +252,7 @@ public static DataBlock readFrom(ByteBuffer buffer)
       throws IOException {
     List<DataBuffer> dataBuffers = buffers.stream()
-        .map(PinotByteBuffer::wrap)
+        .map(PinotByteBuffer::slice)
```
Because PinotByteBuffer.wrap assumes the buffer will be stable and their cursors (position, limit, etc) won't be changed. When using wrap, tests fail. Probably this situation doesn't happen in real live, where byte buffers are going to be sent to other JVMs, but it looks like we are changing buffers during tests. Anyway, it is safer to slice here to be sure these cursors are immutable.

yashmayya replied:
Ah, thanks for that clarification, I was checking non-testing code only and wasn't able to find any scenarios where the position / limit would be modified.

Jackie-Jiang replied:
This is adding overhead. Can you share which test is failing because of this?

gortiz commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/channel/MailboxContentObserver.java:
```diff
@@ -60,10 +63,15 @@ public void onNext(MailboxContent mailboxContent) {
     }
+    _mailboxBuffers.add(mailboxContent.getPayload().asReadOnlyByteBuffer());
+    if (mailboxContent.getWaitForMore()) {
+      return;
```
A debug may not hurt, but it won't be advantageous either. In production, this debug will be disabled, and GRPC guarantees order and arrival. Therefore, if a block split is missed, we will encounter other exceptions.

gortiz commented:
>  the BlockSplitter used in BlockExchange to split blocks based on rows doesn't serve any other purpose than trying to keep the gRPC message size under the max limit right? Any reason we can't get rid of that with this new strategy immediately? I'm not sure if I'm missing anything but it seems like unnecessary overhead now?

Good catch. It is pretty useless and the plan is to remove it in the future. But we decided to keep it for now to not affect upgrades. I mean, during an upgrade updated servers will split with BlockSplitter to make sure not upgraded servers do not receive blocks that are larger than they use to be.

@albertobastos this reminds me that one thing we forgot is to add the flag that enables/disables this feature. We need to disable wait for more (always send complete blocks) during an upgrade.

albertobastos commented on pinot-query-runtime/src/test/java/org/apache/pinot/query/mailbox/MailboxServiceTest.java:
```diff
@@ -87,7 +87,7 @@ public void testLocalHappyPathSendFirst()
         _mailboxService1.getSendingMailbox("localhost", _mailboxService1.getPort(), mailboxId, Long.MAX_VALUE, _stats);
     for (int i = 0; i < ReceivingMailbox.DEFAULT_MAX_PENDING_BLOCKS - 1; i++) {
-      sendingMailbox.send(OperatorTestUtil.block(DATA_SCHEMA, new Object[]{i}));
+      sendingMailbox.send(OperatorTestUtil.block(DATA_SCHEMA, new Object[]{Integer.toString(i)}));
```
In `testRemoteBufferFull` we need to be sure about the payload size to avoid splitting, and that's easier when using strings instead of ints in the schema. So we changed the `DATA_SCHEMA` constant to be `{STRING}` instead of `{INT}`. Doing that, we need to change the rest of tests to fill the block with strings instead of ints to avoid casting errors.

albertobastos commented:
> @albertobastos this reminds me that one thing we forgot is to add the flag that enables/disables this feature. We need to disable wait for more (always send complete blocks) during an upgrade.

You're right @gortiz, the flag was missing. Added a `pinot.query.runner.enable.data.block.payload.split` configuration flag to disable the block splitting during the upgrade process. It defaults to `true`.

gortiz commented on pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:
```diff
@@ -1508,6 +1508,13 @@ public static class MultiStageQueryRunner {
+     */
+    public static final String KEY_OF_ENABLE_DATA_BLOCK_PAYLOAD_SPLIT =
+          "pinot.query.runner.enable.data.block.payload.split";
+    public static final boolean DEFAULT_ENABLE_DATA_BLOCK_PAYLOAD_SPLIT = true;
```
It should be false to simplify the upgrade. It can be changed to true after the next release.

albertobastos replied:
Ok done.

Jackie-Jiang commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/GrpcSendingMailbox.java:
```diff
@@ -52,24 +59,36 @@
     _port = port;
     _deadlineMs = deadlineMs;
     _statMap = statMap;
+    _splitBlocks = config.getProperty(
```
We should move the config parsing logic into `MailboxService` constructor to avoid parsing config on a per mailbox level

Jackie-Jiang commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/GrpcSendingMailbox.java:
```diff
@@ -156,19 +176,30 @@ private StreamObserver<MailboxContent> getContentObserver() {
+      } else {
+        byteStrings = List.of(DataBlockUtils.toByteString(dataBlock));
+      }
+      int sizeInBytes = byteStrings.stream().mapToInt(ByteString::size).reduce(0, Integer::sum);
```
I'd avoid using functional API at per block level.

Jackie-Jiang commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/GrpcSendingMailbox.java:
```diff
@@ -156,19 +176,30 @@ private StreamObserver<MailboxContent> getContentObserver() {
-      ByteString byteString = DataBlockUtils.toByteString(dataBlock);
-      int sizeInBytes = byteString.size();
+      List<ByteString> byteStrings;
+      if (_splitBlocks) {
```
Consider moving this check one level up (into `sendInternal()`) to separate the handling of split vs non-split to reduce overhead of non-split case

Jackie-Jiang commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/channel/MailboxContentObserver.java:
```diff
@@ -47,11 +48,13 @@ public class MailboxContentObserver implements StreamObserver<MailboxContent> {
   public MailboxContentObserver(MailboxService mailboxService, StreamObserver<MailboxStatus> responseObserver) {
     _mailboxService = mailboxService;
     _responseObserver = responseObserver;
+    _mailboxBuffers = new ArrayList<>();
```
(MAJOR) Is this thread safe? Are we sharing the same observer across multiple queries?

albertobastos replied:
As we discussed offline, Mailbox Observer is not shared among concurrent queries and therefore is safe to just stack blocks until a block with `waitForMore=false` is received.

Anyway, we will add an integration test with a small buffer and concurrent queries to verify that.

Jackie-Jiang reviewed (commented):
I didn't get a chance to review it before it got merged, but this is one MAJOR comment about thread safety. @albertobastos @gortiz @yashmayya 
Also, at data block level, performance matters. Let's try to reduce overhead

albertobastos reviewed (commented):
Thanks for your comments @Jackie-Jiang ,

Please find next a new PR addressing them and, most importantly, adding a new integration test to ensure thread safety:

https://github.com/apache/pinot/pull/15694
